src/Microsoft.PowerShell.Commands.Utility/commands/utility/MarkdownOptionCommands.cs

```diff
@@ -256,7 +257,61 @@ public class GetMarkdownOptionCommand : PSCmdlet
         /// </summary>
         protected override void EndProcessing()
         {
-            WriteObject(this.CommandInfo.Module.SessionState.PSVariable.GetValue(MarkdownOptionInfoVariableName, new PSMarkdownOptionInfo()));
+            WriteObject(PSMarkdownOptionInfoCache.Get(this.CommandInfo));
+        }
+    }
+
+    /// <summary>
+    /// The class manages whether we should use a module scope variable or concurrent dictionary for storing the set PSMarkdownOptions.
+    /// When we have a moduleInfo available we use the module scope variable.
+    /// In case of built-in modules, they are loaded as snapins when we are hosting PowerShell.
+    /// We use runspace Id as the key for the concurrent dictionary to have the functionality of separate settings per runspace.
+    /// Force loading the module does not unload the nested modules and hence we cannot use IModuleAssemblyCleanup to remove items from the dictionary.
+    /// Because of these reason, we continue using module scope variable when moduleInfo is available.
+    /// </summary>
+    internal static class PSMarkdownOptionInfoCache
+    {
+        private static ConcurrentDictionary<Guid, PSMarkdownOptionInfo> markdownOptionInfoCache;
+        private const string MarkdownOptionInfoVariableName = "PSMarkdownOptionInfo";
+
+        static PSMarkdownOptionInfoCache()
+        {
+            markdownOptionInfoCache = new ConcurrentDictionary<Guid, PSMarkdownOptionInfo>();
+        }
+
+        internal static PSMarkdownOptionInfo Get(CommandInfo command)
+        {
+            // If we have the moduleInfo then store are module scope variable
+            if (command.Module != null)
+            {
+                return command.Module.SessionState.PSVariable.GetValue(MarkdownOptionInfoVariableName, new PSMarkdownOptionInfo()) as PSMarkdownOptionInfo;
+            }
+
+            // If we don't have a moduleInfo, like in PowerShell hosting scenarios, use a concurrent dictionary.
+            if (markdownOptionInfoCache.TryGetValue(Runspace.DefaultRunspace.InstanceId, out PSMarkdownOptionInfo cachedOption))
+            {
+                // return the cached options for the runspaceId
+                return cachedOption;
+            }
+            else
+            {
+                // no option cache so cache and return the default PSMarkdownOptionInfo
+                var newOptionInfo = new PSMarkdownOptionInfo();
+                return markdownOptionInfoCache.GetOrAdd(Runspace.DefaultRunspace.InstanceId, (key) => newOptionInfo);
```

Comment:
You can use the overload `GetOrAdd(TKey, TValue)`, so no need to create a delegate.

Reply:
Fixed.